File: 1_PythonServer/app/domain/llm/provider_config.py
# 供应商配置领域模型
# 保存的用户配置（API 密钥列表、自定义基础 URL 等）

from collections.abc import Mapping
from dataclasses import dataclass, field

from app.domain.llm.reasoning_replay import ReasoningReplayMode


@dataclass(frozen=True, slots=True)
class ProviderApiKeyConfig:
    """单条 API 密钥配置：密钥引用、提示（脱敏）、轮询权重"""

    key_id: str
    provider_id: str
    api_key_hint: str | None
    poll_weight: int
    sort_order: int
    created_at: str
    updated_at: str
    api_key_ciphertext: str | None = None


@dataclass(frozen=True, slots=True)
class ProviderConfig:
    """供应商配置：完整生成地址、模型发现地址和 API 密钥列表。"""

    provider_id: str
    api_base_url: str
    enabled: bool
    api_keys: tuple[ProviderApiKeyConfig, ...]
    created_at: str
    updated_at: str
    model_discovery_url: str | None = None
    protocol_family: str | None = None
    generation_urls: Mapping[str, str] = field(default_factory=dict)
    generation_auth_schemes: Mapping[str, str] = field(default_factory=dict)
    model_discovery_strategy: str | None = None
    model_discovery_auth_scheme: str | None = None
    updated_generation_protocol: str | None = None
    reasoning_replay_mode: ReasoningReplayMode = ReasoningReplayMode.TOOL_CALL_ROUNDS

File: 2_ReactWeb/src/entities/llm-provider/model/providerConfig.ts
import type {
  ProviderAuthScheme,
  ProviderModelDiscoveryStrategy,
  ProviderProtocolFamily,
} from "./providerCatalog";

export type ProviderReasoningReplayMode = "never" | "tool_call_rounds" | "always";

export interface ProviderApiKeyConfig {
  key_id: string;
  has_api_key: boolean;
  api_key_hint: string | null;
  poll_weight: number;
  rpm: number;
}

export interface ProviderConfig {
  provider_id: string;
  api_base_url: string;
  protocol_family: ProviderProtocolFamily;
  generation_urls: Partial<Record<ProviderProtocolFamily, string>>;
  auth_scheme: ProviderAuthScheme;
  generation_auth_schemes: Partial<Record<ProviderProtocolFamily, ProviderAuthScheme>>;
  model_discovery_url: string | null;
  model_discovery_strategy: ProviderModelDiscoveryStrategy;
  model_discovery_auth_scheme: ProviderAuthScheme;
  enabled: boolean;
  prompt_cache_retention_seconds: number;
  reasoning_replay_mode: ProviderReasoningReplayMode;
  api_keys: ProviderApiKeyConfig[];
  created_at: string;
  updated_at: string;
}

export interface ProviderConfigListResponse {
  count: number;
  items: ProviderConfig[];
}

export interface ProviderModelCheckResult {
  provider_id: string;
  model_id: string;
  ok: boolean;
  checked_url: string;
  selected_key_id?: string | null;
  selected_api_key_hint?: string | null;
}

export interface ProviderApiKeyConfigSaveInput {
  key_id?: string;
  api_key?: string;
  poll_weight: number;
}

export interface ProviderConfigSaveRequest {
  api_base_url: string;
  protocol_family: ProviderProtocolFamily;
  auth_scheme: ProviderAuthScheme;
  model_discovery_url: string | null;
  model_discovery_strategy: ProviderModelDiscoveryStrategy;
  model_discovery_auth_scheme: ProviderAuthScheme;
  enabled: boolean;
  api_keys: ProviderApiKeyConfigSaveInput[];
  reasoning_replay_mode: ProviderReasoningReplayMode;
}

export interface ProviderPromptCachePolicyResponse {
  provider_id: string;
  prompt_cache_retention_seconds: number;
}
